components/director/internal/domain/tenant/service.go

```diff
@@ -164,20 +175,17 @@ func (s *labeledService) CreateManyIfNotExists(ctx context.Context, tenantInputs
 	return nil
 }
 
-func (s *labeledService) createIfNotExists(ctx context.Context, tenant model.BusinessTenantMapping, subdomain, region string) (string, error) {
-	tenantFromDB, err := s.tenantMappingRepo.GetByExternalTenant(ctx, tenant.ExternalTenant)
-	if err != nil && !apperrors.IsNotFoundError(err) {
-		return "", errors.Wrapf(err, "while checking the existence of tenant with external ID %s", tenant.ExternalTenant)
-	}
-	if tenantFromDB != nil {
-		return tenantFromDB.ID, s.upsertLabels(ctx, tenantFromDB.ID, subdomain, region)
+func (s *labeledService) createIfNotExists(ctx context.Context, tenant model.BusinessTenantMapping, subdomain, region string, action func(context.Context, model.BusinessTenantMapping) error) (string, error) {
+	if err := action(ctx, tenant); err != nil {
+		return "", err
 	}
 
-	if err = s.tenantMappingRepo.Create(ctx, tenant); err != nil {
-		return "", errors.Wrapf(err, "while creating tenant with ID %s and external ID %s", tenant.ID, tenant.ExternalTenant)
+	tenantFromDB, err := s.tenantMappingRepo.GetByExternalTenant(ctx, tenant.ExternalTenant)
+	if err != nil && !apperrors.IsNotFoundError(err) {
```

Comment:
Should return error on not found as well, because otherwise, you will get NPE in the line below.

Also, it is not expected to have NotFound right after insertion, such thing should not be ignored.